crates/matrix-sdk/src/notification_settings/mod.rs

```diff
@@ -12,3 +26,751 @@ pub enum RoomNotificationMode {
+
+    #[async_test]
+    async fn test_unmute_room_not_muted() {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri())).await;
+        let room_id = get_test_room_id();
+        let notification_settings = client.notification_settings().await;
+        let mut rules = notification_settings.rules().await;
+
+        // Initialize with a `MentionsAndKeywordsOnly` mode
+        _ = rules.insert_new_rule(RuleKind::Room, &room_id, false, false).unwrap();
+        notification_settings.set_ruleset(&rules.ruleset).await;
+
+        notification_settings.unmute_room(&room_id, true, 2).await.unwrap();
+
+        // The ruleset must not be modified
+        let room_rules = notification_settings.get_custom_rules_for_room(&room_id).await;
+        assert_eq!(room_rules.len(), 1);
+        assert_matches!(rules.ruleset.get(RuleKind::Room, &room_id),
+            Some(AnyPushRuleRef::Room(rule)) => {
+                assert_eq!(rule.rule_id, room_id);
+                assert!(rule.actions.is_empty());
+            }
+        );
+    }
+
+    #[async_test]
+    async fn test_unmute_room() {
+        let server = MockServer::start().await;
+        Mock::given(method("PUT")).respond_with(ResponseTemplate::new(200)).mount(&server).await;
+        Mock::given(method("DELETE")).respond_with(ResponseTemplate::new(200)).mount(&server).await;
+        let client = logged_in_client(Some(server.uri())).await;
+        let room_id = get_test_room_id();
+        let notification_settings = client.notification_settings().await;
+
+        // Start with the room muted
+        notification_settings
+            .set_room_notification_mode(&room_id, RoomNotificationMode::Mute)
+            .await
+            .unwrap();
+        assert_eq!(
+            Some(RoomNotificationMode::Mute),
+            notification_settings.get_user_defined_room_notification_mode(&room_id).await
+        );
+
+        // Unmute the room
+        notification_settings.unmute_room(&room_id, false, 2).await.unwrap();
+
+        // The user defined mode must have been removed
+        assert!(notification_settings
+            .get_user_defined_room_notification_mode(&room_id)
+            .await
+            .is_none());
+    }
+
+    #[async_test]
+    async fn test_unmute_room_default_mode() {
+        let server = MockServer::start().await;
+        Mock::given(method("PUT")).respond_with(ResponseTemplate::new(200)).mount(&server).await;
+        let client = logged_in_client(Some(server.uri())).await;
+        let room_id = get_test_room_id();
+        let notification_settings = client.notification_settings().await;
+
+        // Unmute the room
+        notification_settings.unmute_room(&room_id, false, 2).await.unwrap();
+
+        // The new mode must be `AllMessages`
+        assert_eq!(
+            Some(RoomNotificationMode::AllMessages),
+            notification_settings.get_user_defined_room_notification_mode(&room_id).await
```

Comment:
(One person likes yoda conditions here :eyes: :grin: )

```suggestion
            notification_settings.get_user_defined_room_notification_mode(&room_id).await
            Some(RoomNotificationMode::AllMessages),
```